### PR TITLE
feat(tools): make tools composable

### PR DESCRIPTION
**Pull request by Tomas2D**
Contributes to #166

### CODE REVIEW

File: examples/tools/similarity.ts
Comment (Tomas2D) [inline examples/tools/similarity.ts:1]: `toTool` - mapping new schema to existing tool input
`fromTool` - mapping tool output to similarity tool input

File: src/tools/similarity.ts [hunk 1]
```diff
@@ -59,39 +75,62 @@ export class SimilarityTool<TProviderOptions> extends Tool<
...
-      "desc",
-    ]);
-    const filteredResultsWithDocumentIndices = sortedResultsWithDocumentIndices.slice(
-      0,
-      options?.maxResults ?? this.options.maxResults,
-    );
+    return (config: {
+      toTool: (input: z.output<S>) => ToolInput<T>;
+      fromTool: (
+        input: z.output<S>,
```
Comment (michael-desmond) [inline src/tools/similarity.ts:85]: Given that SimilarityTool is also a tool, it may be clearer to name these `toWrappedTool` and `fromWrappedTool`

**Review by michael-desmond — APPROVED** [review]

File: src/tools/similarity.ts
Comment (Tomas2D) [inline src/tools/similarity.ts:1]: refactor, functionality preserved

File: src/tools/base.ts [hunk 2]
```diff
@@ -383,6 +384,130 @@ export abstract class Tool<
   loadSnapshot(snapshot: ToolSnapshot<TOutput, TOptions>): void {
     Object.assign(this, snapshot);
   }
+
+  fork<S extends AnyTool, TS extends ZodSchema>(
```
[the remark this answers is not in the document]
Comment (Tomas2D) [reply src/tools/base.ts:388]: They are not needed anymore. Removed.

**Review by jezekra1 — APPROVED** [review]

**Pull request merged**